crates/orderbook/src/api/get_app_data.rs

```diff
@@ -0,0 +1,43 @@
+use {
+    crate::database::Postgres,
+    anyhow::Result,
+    model::app_id::AppDataHash,
+    reqwest::StatusCode,
+    std::convert::Infallible,
+    warp::{
+        reply::{with_header, with_status},
+        Filter,
+        Rejection,
+        Reply,
+    },
+};
+
+pub fn request() -> impl Filter<Extract = (AppDataHash,), Error = Rejection> + Clone {
+    warp::path!("v1" / "app_data" / "by_hash" / AppDataHash).and(warp::get())
+}
+
+pub fn get(
+    database: Postgres,
+) -> impl Filter<Extract = (Box<dyn Reply>,), Error = Rejection> + Clone {
+    request().and_then(move |contract_app_data: AppDataHash| {
+        let database = database.clone();
+        async move {
+            let result = database.get_full_app_data(&contract_app_data).await;
+            Result::<_, Infallible>::Ok(match result {
+                Ok(Some(response)) => {
+                    let response = with_status(response, StatusCode::OK);
+                    let response = with_header(response, "Content-Type", "application/json");
```

Comment:
Is the header `Content-Type application/json` still correct if you only return a string?
If not, `Content-Type: text/html; charset=utf-8` would probably be desired.

Reply:
Good point. The more I think of it, I don't think we should limit app-data to being JSON. For example, in the `contracts` repo we use this for fee withdrawals:

https://github.com/cowprotocol/contracts/blob/5c25837f585d79dc5450fb80f66c41f596c3198d/src/tasks/dump.ts#L61

This would be a small change to the specification, but it would require changes to:

https://github.com/cowprotocol/services/blob/615cb7cc1ee7f45a35670948868810032882d9d2/crates/shared/src/app_data.rs#L38-L46

Where it would accept anything as long as the maximum size is respected and just return a `Default::default()` if it fails to parse JSON or the field is missing.

Reply:
The header is correct because we enforce that app data is json. If we don't want to enforce that then indeed the content type shouldn't be json. If you want to be as general as possible then it should not even be string. It should be raw bytes.

To me it makes sense to enforce json. Our backend app data scheme is (going to be) defined on json. Why would we accept app data that doesn't match that? Imagine you allow arbitrary bytes or strings and an api user sets a document that is supposed to contain backend app but fails serde_json deserialization for some reason. Maybe a typo making it invalid json or maybe a number that is too large for u64 (assuming serde_json would error in this case). Then we would silently accept this and internally treat it as default app data.

Reply:
You have a good point. I agree that enforcing JSON to avoid API usage error makes sense (and this is the intended usage anyway). Also, this is what we specified initially, and I am derailing a bit 😛.

In the future we may also want to support a "Solidity friendly" app-data specification so that it can be computed on-chain (which would likely be EIP-712 struct hashing based, so "raw bytes"), but this is out of scope for now.